case_study/memory_leaks/images_1/lib/main.dart

```diff
@@ -0,0 +1,268 @@
+
+import 'package:flutter/material.dart';
+
+void main() {
+  runApp(MyApp());
+}
+
+class MyApp extends StatelessWidget {
+  // This widget is the root of your application.
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      title: 'Flutter Demo',
+      theme: ThemeData(
+        primarySwatch: Colors.blue,
+      ),
+      home: const MyHomePage(
+        title: 'Flutter Memory Case Study #1',
+      ),
+      // showPerformanceOverlay: true,
+    );
+  }
+}
+
+class MyHomePage extends StatefulWidget {
+  const MyHomePage({Key key, this.title}) : super(key: key);
+
+  final String title;
+
+  @override
+  MyHomePageState createState() => MyHomePageState();
+}
+
+class MyHomePageState extends State<MyHomePage>
+    with SingleTickerProviderStateMixin {
+  TabController _tabController;
+
+  List tabs = ['1', '2', '3'];
+
+  @override
+  void initState() {
+    super.initState();
+    _tabController = TabController(length: tabs.length, vsync: this);
+  }
+
+  Widget recordLoadedImage(ImageChunkEvent imageChunkEvent, String imageUrl) {
+    if (imageChunkEvent == null) return null;
+
+    final recordLoading = loadedImages.putIfAbsent(imageUrl, () {
+      print(
+        'Loading size: ${imageChunkEvent.expectedTotalBytes},'
+        ' image: $imageUrl',
+      );
+      return imageChunkEvent;
+    });
+
+    final expectedTotalBytes = recordLoading.expectedTotalBytes;
+    final cumulativeBytes = recordLoading.cumulativeBytesLoaded;
+
+    final loadingState = cumulativeBytes == expectedTotalBytes
+        ? 'Loaded.'
+        : 'Loading, $cumulativeBytes of $expectedTotalBytes bytes.';
+
+    print('$imageUrl $loadingState');
+
+    // Progress spinnger.
+    return Center(
+      child: CircularProgressIndicator(
+        value: expectedTotalBytes != null
+            ? cumulativeBytes / expectedTotalBytes
+            : null,
+      ),
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    // The build method rerun every time setState is called.
+    //
+    // The Flutter framework has been optimized to efficiently run the build.
+    // So that only what needs updating rather than having to individually
+    // change instances of widgets.
+    final scaffold = Scaffold(
+      backgroundColor: Colors.black,
+      appBar: AppBar(
+        title: const Text('Image Viewer'),
+        actions: <Widget>[
+          IconButton(
+            // action button
+            icon: const Icon(Icons.add),
+            onPressed: () {
+              _jumpNewPage();
+            },
+          )
+        ],
+        bottom: TabBar(
+            controller: _tabController,
+            tabs: tabs.map((e) => Tab(text: e)).toList()),
+      ),
+      body: listView(),
+    );
+
+    return scaffold;
+  }
+
+  /// *********************************************************************************
+  /// ***** ListView outside of TabBarView, ListView w/o width/height smaller.    *****
```

Comment:
Also consider removing the print statements as users won't have the benefit of them in typical apps they are debugging.

Reply:
I meant to change these to developer.log.